### PR TITLE
feat(pie-modal): dsw-975 - add hasBackButton prop

### PR DESCRIPTION
**Pull request by jamieomaguire**
---
"@justeattakeaway/pie-modal": minor
"pie-storybook": minor
---

[Added] - hasBackButton prop to modal

---
"@justeattakeaway/pie-webc-core": minor
---

[Changed] - Improve TS typings in RTL mixin and provde extendable prop interface for components

---
"pie-storybook": minor
---

[Changed] - Use strict typings to ensure storybook build errors when component interfaces change

### CODE REVIEW

File: .changeset/serious-tables-admire.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+"@justeattakeaway/pie-webc-core": minor
+---
+
+[Changed] - Improve TS typings in RTL mixin and provde extendable prop interface for components
```
Comment (xander-marjoram) [inline .changeset/serious-tables-admire.md:5]:
```suggestion
[Changed] - Improve TS typings in RTL mixin and provide extendable prop interface for components
```

File: packages/components/pie-modal/src/index.ts [hunk 2]
```diff
@@ -134,7 +149,13 @@ export class PieModal extends RtlMixin(LitElement) implements ModalProps {
...
         if (previousValue !== undefined) {
             if (previousValue) {
-                this._dispatchModalCustomEvent(ON_MODAL_CLOSE_EVENT);
+                if (this._backButtonClicked) {
+                    // Reset the flag
+                    this._backButtonClicked = false;
+                    this._dispatchModalCustomEvent(ON_MODAL_BACK_EVENT);
+                } else {
+                    this._dispatchModalCustomEvent(ON_MODAL_CLOSE_EVENT);
+                }
```
Comment (ashleynolan) [inline packages/components/pie-modal/src/index.ts:158]: This totally isn't caused from your changes on their own, but I think they just exacerbated my confusion reading it!

I think `previousValue` isn't very clear. It should likely be called something more descriptive. I'm guessing it gets the current (changed) value of `isOpen`, but I'm not entirely sure tbh.

Either a name change, or some comments would definitely help (if you know a bit more about the context here.

**Review by ashleynolan — COMMENTED** [review]
Few comments from me – looks good though 👍

File: apps/pie-storybook/stories/pie-modal.stories.ts [hunk 3]
```diff
@@ -1,45 +1,53 @@
...
 import { html, TemplateResult } from 'lit';
 import { ifDefined } from 'lit/directives/if-defined.js';
-import '@justeattakeaway/pie-button'; // Ensures the button WC is available for use in the templates
-
+import '@justeattakeaway/pie-button'; // Register pie-button
+import '@justeattakeaway/pie-icon-button'; // Register pie-icon-button
+import '@justeattakeaway/pie-modal'; // Register pie-modal
+import '@justeattakeaway/pie-icons-webc/icons/IconClose'; // Register icon-close
+import '@justeattakeaway/pie-icons-webc/icons/IconChevronLeft'; // Register icon-chevron-left
+import '@justeattakeaway/pie-icons-webc/icons/IconChevronRight'; // Register icon-chevron-right
```
[the remark this answers is not in the document]
Comment (jamieomaguire) [reply apps/pie-storybook/stories/pie-modal.stories.ts:9]: I'm unsure. They wouldn't appear unless I did this. Can we catch up later today?

File: apps/pie-storybook/stories/pie-modal.stories.ts [hunk 4]
```diff
@@ -1,45 +1,53 @@
...
+type ModalStoryArgTypes = {
+    [K in keyof ModalProps]: object;
+}
+
+type ModalStoryMeta = Meta & {
+    argTypes: ModalStoryArgTypes,
+    args: ModalProps
+}
+
+const modalStoryMeta: ModalStoryMeta = {
```
[the remark this answers is not in the document]
Comment (jamieomaguire) [reply apps/pie-storybook/stories/pie-modal.stories.ts:40]: yup thats the plan, just keeping the scope of this change small and will update the generator once this has been merged

File: packages/components/pie-modal/src/index.ts (hunk 2 above)
Comment (jamieomaguire) [reply packages/components/pie-modal/src/index.ts:158]: They are the values the properties were _before_ the update cycle. I think this is actually less confusing. How about `oldValue`?

File: packages/components/pie-modal/src/index.ts [hunk 5]
```diff
@@ -39,6 +44,9 @@ export class PieModal extends RtlMixin(LitElement) implements ModalProps {
     @property({ type: Boolean, reflect: true })
     public isDismissible = false;
 
+    @property({ type: Boolean, reflect: true })
```
[the remark this answers is not in the document]
Comment (jamieomaguire) [reply packages/components/pie-modal/src/index.ts:47]: I thought the pattern was established by `isDismissible` being reflected. Do we have a guide around when to reflect?

File: packages/components/pie-modal/src/index.ts (hunk 2 above)
Comment (jamieomaguire) [reply packages/components/pie-modal/src/index.ts:159]: I've added a comment, let me know if its any more useful :)

**Review by xander-marjoram — APPROVED** [review]

**Review by ashleynolan — APPROVED** [review]

**Pull request merged**